### DISCUSSION TITLE
Is dependency resolution against local repository possible when building test repository?

### DISCUSSION THREAD

**Discussion by chadwick00** [khmarbaise/maven-it-extension#293] [Ideas]
We're writing a plugin that has a substantial set of dependencies. When testing the plugin using maven-it-extension, it takes ~ 6 minutes to build the test_method_name/.m2/repository because as far as I can make out, all the dependencies are being downloaded from our remote repository (making dummy changes to the artifacts in ${user.home}/.m2/repository indicates they aren't being pulled into the test repo). Once the repo has been setup, the tests run in a matter of seconds.

Having each test take 6 minutes to run isn't viable for local development, the testing iterations are too long. Is it possible to configure the maven-it-extension to pull dependencies from the local repo when building the test repos? I've tried a few combinations of \<localRepository\> and \<repository\>file://C:/etc\</repository\> tags in ${user.home}/.m2/repository/settings.xml, but this hasn't worked. I'm able to copy a populated test repo across in the filesystem in ~20sec, so I think this would make development viable again.

**Comment by chadwick00** [comment]
It occurs to me as I write the above, that the single test we currently have creates a test_method_name/.m2/repository of ~ 70Mb. We've contemplated refactoring 16 tests to use the maven-it-extension. After a test suite runs we'll have well over 1Gb of test repos on the filesystem that will need to be cleared up.

**Comment by khmarbaise** [comment]
The first thing is using the configured remote repository (your `settings.xml`) and consume the dependencies from there is the default  and it's intentionally to do so to have each IT clearly separated from each other which makes it possible to parallelize them. 

Furthermore If you have a set of dependencies which is needed to be existing in your local repository for your IT you should check the documentation (https://khmarbaise.github.io/maven-it-extension/itf-documentation/usersguide/usersguide.html#_predefined_repository_content_for_tests).

It is also possible to use a single cache for all of your tests (or a set of them) which I can not recommend because it could influence each other... see https://khmarbaise.github.io/maven-it-extension/itf-documentation/usersguide/usersguide.html#_common_maven_cache

**Reply by chadwick00** [reply]
Hi,

Many thanks for the pointers to the caching strategies sections, I hadn't properly digested the content here. I'll give them a whirl.

With regards to using the remote repository as the "default", I still think (happy to be proven wrong) that I'd like to consume dependencies from the local .m2 cache first before going to the remote cache when building the test repositories. Is this possible?

Regards.

**Comment by rakus** [comment]
The invoker plugin sets a property named `localRepositoryUrl`. It is the path to the local repository as file-url (`file://...`). With this it is possible to create a test(s) specific settings.xml that used the local repository as a "remote" repository.

So the running test will first try to "download" an artifact from `~/.m2/repository`. If the artifact is not available there it will download it from one of the other repositories configured.

I tested this with ITF by using a test-local `settings.xml` and created the url to the local repo like the following example. This works on Linux, but I doubt it would work on Windows.

```
      <repositories>
        <repository>
          <id>local.central</id>
          <url>file:@settings.localRepository@</url>
          <releases>
            <enabled>true</enabled>
          </releases>
          <snapshots>
            <enabled>true</enabled>
          </snapshots>
        </repository>
        ....
      </repositories>
```

I think it would be useful if ITF would also provide something like `localRepositoryURL` while copying resources.

Regards,
Ralf

**Reply by rakus** [reply]
BTW: `<url>file:///@settings.localRepository@</url>` (3 slashes) also works on Linux and _might_ work on Windows.
Anyway it would good to have `localRepositoryURL`.
